### PR TITLE
[ocamlformat] Enable for funind.

### PR DESCRIPTION
**Pull request by ejgallego**
As part of the proof refactoring work I am doing some modifications to
`funind` and indentation of that code is driving me a bit crazy; I'd
much prefer to delegate it to an automatic tool.

### CODE REVIEW

File: .ocamlformat-ignore
Comment (Zimmi48) [inline .ocamlformat-ignore]: This extra new line is making the linter fail.

**Review by Zimmi48 — COMMENTED** [review]
Seems sensible to me. Up to @coq/funind-maintainers. However, you might want to let #11767 go in first.

File: .ocamlformat-ignore
Comment (ejgallego) [reply .ocamlformat-ignore]: Linter is failing due to non-matching ocamlformat versions, #11778 should address this.

**Comment by ejgallego** [comment]
> However, you might want to let #11767 go in first.

Sure, all the pending PRs that would touch funind should go before this one.

**Comment by Matafou** [comment]
Ok for me once #11767 is merged.

**Comment by maximedenes** [comment]
@ejgallego I have a question: will the linter fail if funind files are not formatted in a way that conforms with `ocamlformat`, or will it be fixed by the pre-commit hook?

One clear contract we had with the linter is that all the errors it emits should be addressed by the hook.

**Comment by ejgallego** [comment]
> One clear contract we had with the linter is that all the errors it emits should be addressed by the hook.

Good point, indeed if we want to this on the linter is easy, just call in the hook `dune build @fmt --auto-promote`

**Comment by maximedenes** [comment]
> Good point, indeed if we want to this on the hook is easy, just call `dune build @fmt --auto-promote`

Does this work even if we built Coq with `make`? (or if we didn't build Coq at all I imagine, if it is just a matter of calling the formatter)

File: .ocamlformat-ignore
Comment (maximedenes) [reply .ocamlformat-ignore]: Could we switch to a positive listing of files instead of a negative one?

**Comment by ejgallego** [comment]
> Does this work even if we built Coq with `make`? (or if we didn't build Coq at all I imagine, if it is just a matter of calling the formatter)

Unfortunately it won't work if you have a dirty tree, as `dune` needs to setup its build rules before calling the formatter and that includes the promote rules for the output of the formatter.

You could have the pre-commit hook work on a clean checkout, but that seems overkill. I guess people using make meanwhile can follow the recommendations for editor integration,.

**Comment by maximedenes** [comment]
> the recommendations for editor integration

Do you have a pointer? Our own files point to "the docs".

**Comment by ejgallego** [comment]
> Do you have a pointer? Our own files point to "the docs".

https://github.com/ocaml-ppx/ocamlformat#editor-setup

Ideal would be indeed for the hook to take care of it tho.

**Comment by maximedenes** [comment]
There's nothing for VsCode, for example...

I suggest we do the following: for now, we merge only the patch which formats funind. After the switch to Dune, we add the call to the formatter in the hook.

Can you report the issue about the negative listing too? I would prefer to see it fixed before we make more files be formatted automatically.

**Comment by maximedenes** [comment]
> only the patch which formats funind

In particular, we should keep `.ocaml-format.ignore` unchanged for now IMHO.

**Comment by Zimmi48** [comment]
> I suggest we do the following: for now, we merge only the patch which formats funind. After the switch to Dune, we add the call to the formatter in the hook.

OK to me

**Comment by ejgallego** [comment]
> OK to me

That is inconsistent with what we do for micromega, tho?

IMHO I agree with the original reasoning that was done with micromega, not enforcing the formatting kinda defeats the purpose; if the plugin is not included in the list of ocamlformat-enabled targets then `dune build @fmt` will not do anything on it even if called voluntarily.

I dunno what the best tradeoff is here.

**Comment by Zimmi48** [comment]
> That is inconsistent with what we do for micromega, tho?

It is. Indeed, I don't like this solution, but we cannot force the hand of developers too hard either. @maximedenes Can you remind us what is still blocking you from using Dune?

> IMHO I agree with the original reasoning that was done with micromega, not enforcing the formatting kinda defeats the purpose; if the plugin is not included in the list of ocamlformat-enabled targets then `dune build @fmt` will not do anything on it even if called voluntarily.

That's a good point. Then what could be done would be to disable the check in the linter.

**Comment by ejgallego** [comment]
I folks, what do you think of this? It stalled a bit.

Personally I've become quite a fan of OCamlformat [after being ultra opposed to it, funny] maybe because in Coq we have such an inconsistent style that I guess I much prefer consistency over my own personal preferences / code art.

I've rebased the PR in case we'd like to merge; I'm also fine to close this if there is opposition.

**Comment by ejgallego** [comment]
Note that it got @Matafou 's approval so I lean towards merging.

**Comment by Zimmi48** [comment]
Now that the `ocamlformat` infrastructure is in good shape, and given that `ocamlformat` is already activated on micromega, and that the pre-commit hook takes care of fixing improperly formatted code, there does not need to be a general discussion. It's a decision for @coq/funind-maintainers to make. @Matafou could take care of approving + merging this PR.

**Review by Matafou — APPROVED** [review]

**Comment by Matafou** [comment]
merge.sh complains about this PR not run with the latest version of the CI. Should I care?

**Comment by Zimmi48** [comment]
No, that's not an issue. (You can display the diff and it should be pretty small and unrelated IINM)

**Comment by Matafou** [comment]
Merged.
I am not familiar with milestones. Can someone put one please?

**Comment by ejgallego** [comment]
Thanks @matafou , milestones are mainly useful to determine if a change needs backporting.

This doesn't so targeting 8.12 beta should work.

**Pull request merged**
